src/repr/binary_codec.ml

```diff
@@ -0,0 +1,271 @@
+include Binary_codec_intf
+open Staging
+
+let unsafe_add_bytes b k = k (Bytes.unsafe_to_string b)
+let str = Bytes.unsafe_of_string
+
+let charstring_of_code : int -> string =
+  let tbl =
+    Array.init 256 (fun i -> Bytes.unsafe_to_string (Bytes.make 1 (Char.chr i)))
+  in
+  fun [@inline always] i ->
+    assert (i < 256);
+    Array.unsafe_get tbl i
+
+module Unit = struct
+  let encode () _k = ()
+  let decode _ ofs = (ofs, ()) [@@inline always]
+end
+
+module Char = struct
+  let encode c k = k (charstring_of_code (Char.code c))
+  let decode buf ofs = (ofs + 1, buf.[ofs]) [@@inline always]
+end
+
+module Bool = struct
+  let encode b = Char.encode (if b then '\255' else '\000')
+
+  let decode buf ofs =
+    let ofs, c = Char.decode buf ofs in
+    match c with '\000' -> (ofs, false) | _ -> (ofs, true)
+end
+
+module Int8 = struct
+  let encode i k = k (charstring_of_code i)
+
+  let decode buf ofs =
+    let ofs, c = Char.decode buf ofs in
+    (ofs, Stdlib.Char.code c)
+    [@@inline always]
+end
+
+module Int16 = struct
+  let encode i =
+    let b = Bytes.create 2 in
+    Bytes.set_uint16_be b 0 i;
+    unsafe_add_bytes b
+
+  let decode buf ofs = (ofs + 2, Bytes.get_uint16_be (str buf) ofs)
+end
```

Comment:
I was wondering if for these functions where we know exactly the size of the encoding we wouldn't reduce the number of allocations by writing something like:

```ocaml
let encode = 
  let b = Bytes.create 2 in
  fun i ->
    Bytes.set_uint16_be b 0 i;
    unsafe_add_bytes b
```

But it may be useless. I can test it you want

Reply:
Unfortunately, this wouldn't work for this encoder type for two reasons:
- the receiver function `(string -> unit) -> unit` is permitted to keep a reference to the string that it's passed / delay reading from it, which makes the `unsafe_add_bytes` unsound;
- sharing global buffers makes the API non-threadsafe, since concurrent encodes may observe each others' intermediate state.

A fix to this intermediate allocations problem is probably to use your improved encoder type instead :-) (https://github.com/mirage/repr/pull/57)

Reply:
Yep, you're right, thanks :-)